### DISCUSSION TITLE
Use same meta-info part for Java and spark SDK

### DISCUSSION THREAD

**Discussion by Thespica** [apache/incubator-graphar#379] [General]
Java and scala work on JVM and can call each other's code directly. We can let Java SDK and spark SDK use the same meta-info part so
that we can avoid duplicate work when updating.

For the meta-info part, there are two options:
- Rewrite by Java, but may affect spark SDK's existing code
- Let Java use meta-info written in scala.

I'm not sure about how to choose.

Related issue: #276

**Comment by SemyonSinchenko** [comment]
I think that scala-info objects are not the better choice just because of scala. For PySpark part calling Java code better than Scala too (py4j is not well suited for scala).

I suggest have:
- Java submodule with Info meta-objetcs and serialization/deserialization (need to think how to work with S3/HDFS yaml files)
- Java submodule with writter/reader logic that depends on Java-meta
- Scala Spark datasources submodules (spark 32, spark 33, etc.)
- Scala Spark helpers submodule (read/write/index) that depends on Java meta and chosen datasource implementation for a specific version of spark
- pyspark bindings to both java meta and scala helpers

**Reply by SemyonSinchenko** [reply]
It should also simplify testsing and increase the CI-stages speed

**Comment by acezen** [comment]
Agree to Sem, The meta-part is the core and implementation of GraphAr format, unified with Java submodule would help to keep the format implementation alignment between libraries.

**Comment by Thespica** [comment]
Which naming convention should we use, remain existing style or use normal Java naming convention？

existing style:
```
bool is_nullable;
bool getIs_nullable();
void setIs_nullable(bool is_nullable);
```

normal naming convention in Java:

```
bool nullable;
bool isNullable();
void setNullable(bool nullable);
```

[bool naming convention in java](https://stackoverflow.com/questions/5322648/for-a-boolean-field-what-is-the-naming-convention-for-its-getter-setter)

**Reply by acezen** [reply]
I suggest we use the Java naming convention. It's good to read and follow the best practice.

you can refer to the convention of [google java style](https://google.github.io/styleguide/javaguide.html)

**Comment by SemyonSinchenko** [comment]
I see it like this:
![2024-03-12-Note-21-51](https://github.com/alibaba/GraphAr/assets/29755009/ea6fdd84-a487-431a-b11d-520548402458)

We have:
- top-level Java Info implementations
- standalone Java read-write (via java parquet/orc libs) depends of Java Info
- spark datasources: ds-32, ds-33, etc.
- spark Read/Write/GenerateIndex/Util depends of Java Info AND of one of datasources
- python/pyspark provide `py4j` free implementation of Info AND pure PySpark Read/Write/Index/Util that depends of one of datasources

With this case PySpark implementation will work on both regular PySpark and SparkConnect jobs. And also there is an option to use pure java (with provided Info) without Spark (do we really need it?)

@Thespica @acezen FYI as a continue of the discussion in #401

**Reply by acezen** [reply]
Fantastic work！thanks Sem. About the pure java without Spark option, yes, there is  requirement like [HugeGraph](https://github.com/apache/incubator-hugegraph) that need a JAVA read/write to process the GraphAr data.

**Comment by Thespica** [comment]
A point needs to be discussed here about immutability.

We hope meta-info is immutable for ability of parallel and safety, but the `final` keep word is not enough to contract that, it only ensure
the const of pointer, rather than the whole object.

So we need some immutable class to replace original collection like List, Map and etc. Here are two options I found:

1. use `UnmodifiableXxx`(i.e `UnmodifiableMap`) from java collection. It also can be modified from original map, but we can avoid by copying as [they discussed](https://stackoverflow.com/a/22636750/21205409). 
2. use `ImmutableXxx`(i.e `ImmutableMap`) from guava. It's really immutable, but we need add a new dependency.

I prefer the first now, but I'm not sure about that, so I need your perspective.

Besides, we should return  `UnmodifiableXxx` or `ImmutableXxx` directly rather than returning interface(i.e `Map` or `List`) to hint user that the object cannot be modified.

**Reply by SemyonSinchenko** [reply]
Option one looks better for me

**Reply by acezen** [reply]
Option one looks good to me too.
